### PR TITLE
Fix REPLACE PARTITION waiting for mutations/merges on unrelated partitions

### PR DESCRIPTION
**Pull request by Enmk**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
`ALTER TABLE .. REPLACE PARTITION` doesn't wait anymore for mutations/merges that happen in other partitions

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.
On a big table with lots of parts/partitions and lots of merges/mutations scheduled on a table, `ALTER TABLE .. REPLACE PARTITION` waits very long until all of those finish, even if the target partition is not affected by those merges and/or mutations. This leads to poor maintenance experience.

With this improvement, `ALTER TABLE .. REPLACE PARTITION` only waits for merges/mutations to the target partition, greatly reducing up the waiting period.

Technically this stops the merges on all partitions and prevents new ones from starting, rather than stopping merges on the target partition only. But since `REPLACE` is relatively simple, there is a high chance that merges and mutations are re-allowed relatively fast.

Related: #45328

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
This is an automated comment for commit 35846083f994c8f95fae610830f5c01eff945a81 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/54272/35846083f994c8f95fae610830f5c01eff945a81/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🔴 failure</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Storages/MergeTree/MutateTask.cpp [hunk 1]
```diff
@@ -881,11 +872,22 @@ struct MutationContext
     bool need_prefix = true;
 
     scope_guard temporary_directory_lock;
+
+    bool checkOperationIsNotCanceled() const
+    {
+        if (merges_blocker->isCancelledForPartition(new_data_part->info.partition_id)
```
Comment (Enmk) [inline src/Storages/MergeTree/MutateTask.cpp:878]: I assume here that the merge operation doesn't create parts in other partitions, so we might check only if the destination partition was canceled or not.

File: src/Storages/MergeTree/MergeTask.cpp [hunk 2]
```diff
@@ -122,6 +122,20 @@ static void addMissedColumnsToSerializationInfos(
     }
 }
 
+bool MergeTask::GlobalRuntimeContext::isCancelled() const
+{
+    return merges_blocker->isCancelledForPartition(future_part->part_info.partition_id)
```
Comment (Enmk) [inline src/Storages/MergeTree/MergeTask.cpp:127]: I assume here that the merge operation doesn't create parts in other partitions, so we might check only if the destination partition was canceled or not.

**Comment by alexey-milovidov** [comment]
It failed with a segmentation fault because your code is not memory-safe.

**Pull request closed without merging**